Source/MQTTnet/Server/AsyncQueue.cs

```diff
@@ -0,0 +1,114 @@
+﻿using MQTTnet.Internal;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace MQTTnet.Server
+{
+    public class AsyncQueue<T> : IDisposable
+    {
+        private readonly Queue<T> _queue = new Queue<T>();
+        private readonly AsyncAutoResetEvent _queueEmptyEvent = new AsyncAutoResetEvent();
+        private readonly AsyncAutoResetEvent _queueNotFullEvent = new AsyncAutoResetEvent();
+        protected readonly IMqttServerOptions _options;
+
+
+        public AsyncQueue(IMqttServerOptions options)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public int Count
+        {
+            get
+            {
+                lock (_queue)
+                {
+                    return _queue.Count;
+                }
+            }
+        }
+        
+
+        public void Clear()
+        {
+            lock (_queue)
+            {
+                _queue.Clear();
+            }
+        }
+
+        public void Dispose()
+        {
+        }
+
+        public void Enqueue(T packet)
+        {
+            if (packet == null) throw new ArgumentNullException(nameof(packet));
+
+            TryEnqueue(packet);
+        }
+
+        protected virtual bool TryEnqueue(T packet)
+        {
+            bool wait = false;
+            lock (_queue)
+            {
+                if (_queue.Count >= _options.MaxPendingMessagesPerClient)
+                {
+                    switch (_options.PendingMessagesOverflowStrategy)
+                    {
+                        case MqttPendingMessagesOverflowStrategy.DropNewMessage:
+                            return false;
+                        case MqttPendingMessagesOverflowStrategy.DropOldestQueuedMessage:
+                            _queue.Dequeue();
+                            break;
+                        case MqttPendingMessagesOverflowStrategy.Block:
+                            wait = true;
+                            break;
+                    }
+                }
+
+                _queue.Enqueue(packet);
```

Comment:
Is this intentional? If PendingMessagesOverflowStrategy is MqttPendingMessagesOverflowStrategy.Block, then it's possible to Add the item twice. (Line 73 and 79)

Reply:
nope it is not